dstep/translator/Context.d

```diff
@@ -136,16 +137,120 @@ class Context
         return typedefIndex_.typedefParent(cursor);
     }
 
+    private string translateSpellingImpl(in Cursor cursor)
+    {
+        return cursor.spelling == ""
+            ? generateAnonymousName(cursor)
+            : cursor.spelling;
+    }
+
     public string translateSpelling(in Cursor cursor)
     {
-        auto typedefp = typedefParent(cursor.canonical);
+        if (auto spelling = (cursor in translatedSpellings))
+        {
+            return *spelling;
+        }
+        else
+        {
+            auto spelling = translateSpellingImpl(cursor);
+            translatedSpellings[cursor] = spelling;
+            return spelling;
+        }
+    }
 
-        if (typedefp.isValid && cursor.spelling == "")
-            return typedefp.spelling;
+    private void printCollisionWarning(
+        string spelling,
+        Cursor cursor,
+        Cursor collision)
+    {
+        import std.format : format;
+        import std.stdio : writeln;
+
+        if (options.printDiagnostics)
+        {
+            auto cursorLocation = cursor.location;
+            auto collisionLocation = collision.location;
+
+            auto message = format(
+                "%s:%d:%d: warning: a type renamed to '%s' due to the "
+                "collision with the symbol declared in %s:%d:%d",
+                cursorLocation.path,
+                cursorLocation.line,
+                cursorLocation.column,
+                spelling,
+                collisionLocation.path,
+                collisionLocation.line,
+                collisionLocation.column);
+
+            writeln(message);
+        }
+    }
+
+    private void throwCollisionError(
+        string spelling,
+        Cursor cursor,
+        Cursor collision) {
+        import std.format : format;
+
+        auto cursorLocation = cursor.location;
+        auto collisionLocation = collision.location;
+
+        throw new TranslationException(
+            format(
+                "%s:%d:%d: error: a type name '%s' "
+                "collides with the symbol declared in %s:%d:%d",
+                cursorLocation.path,
+                cursorLocation.line,
+                cursorLocation.column,
+                spelling,
+                collisionLocation.path,
+                collisionLocation.line,
+                collisionLocation.column));
```

Comment:
Perhaps move the formatting of the error message to a separate function since it's the same as for the warning.